src/utils/proxySet.ts

```diff
@@ -0,0 +1,80 @@
+import { CUSTOM_TYPE } from '../vanilla'
+
+// properies that we don't want to expose to the end-user
+interface InternalProxySet<T> extends Set<T> {
+  data: T[]
+  // used for JSON.stringify
+  toJSON(): T[]
+  [CUSTOM_TYPE]: boolean
+}
+
+export function proxySet<T>(initialValues: Iterable<T> = []): Set<T> {
+  const set: InternalProxySet<T> = {
+    data: [...new Set(initialValues)],
+    has(value) {
+      return this.data.indexOf(value) !== -1
+    },
+    add(value) {
+      if (this.data.indexOf(value) === -1) {
+        this.data.push(value)
+      }
+      return this
+    },
+    delete(value) {
+      if (this.data.indexOf(value) === -1) {
+        return false
+      }
+      this.data = this.data.filter((v: T) => v !== value)
+      return true
+    },
+    clear() {
+      this.data = []
+    },
+    get size() {
+      return this.data.length
+    },
+    forEach(cb) {
+      for (let i = 0; i < this.data.length; i++) {
+        const value = this.data[i] as T
+        cb(value, value, this)
+      }
+    },
+    get [Symbol.toStringTag]() {
+      return 'Set'
+    },
+    [Symbol.iterator]() {
+      let index = 0
+
+      return {
+        next: () =>
+          index < this.data.length
+            ? { value: this.data[index++], done: false }
+            : { done: true },
+      } as IterableIterator<T>
+    },
+    [CUSTOM_TYPE]: true,
+    toJSON() {
+      return this.data
+    },
+    values() {
+      return this.data.values()
+    },
+    keys() {
+      // for Set same has Set.values()
+      return this.data.values()
+    },
+    entries() {
+      // array.entries returns [index, value] while Set [value, value]
+      return new Set(this.data).entries()
+    },
+  }
+  Object.defineProperties(set, {
+    data: {
+      writable: true,
+    },
+  })
+
+  Object.seal(set)
+
+  return set as Set<T>
```

Comment:
This should return a proxy in my opinion:
```suggestion
  return proxy(set) as Set<T>
```

Reply:
Just for my understanding. If it returns a proxy this could be used as follow
```js
const state = proxySet([1,2,3])
```
Which is equivalent to
```js
const state = proxy(proxySet([1,2,3]))
```
So returning a proxy as you suggested would allow the end-user to use the `proxySet` directly. Which is more elegant, make sense.

Reply:
Actually, both are considerable.
```js
const state = proxySet(new Set([1, 2, 3])) // returns a proxy
const state = proxy(emulateSet(new Set([1, 2, 3]))
```
I just feel the name `proxySet` takes a `Set` (or any Iterable) and returns a set. If `emulateSet` (or some other name) has some use cases, it can be a good option too. We need to decide which to go.

Reply:
I like your first suggestion, the more I think about it the more It feels natural to me.

```js
const state = proxySet(new Set([1, 2, 3])) // returns a proxy
```
Also, I would say this it's less error-prone given it returns a `proxy`. 

With the latter, I think a user can forget to wrap it with `proxy`.

Reply:
Cool. And we can still do the original usage:
```js
const state = proxy({
  set1: proxySet(new Set([1, 2, 3])),
  set2: proxySet([4, 5, 6])),
})
```